# <font size=35 color=lightgreen>**Sentiment API**<font>🥲
---

## <font size=6 color=#00FFFF>Configuración Inicial (Librerías)</font>

#### 1. Procesamiento y Manipulación de Datos
* **`pandas`**
    * Nos ayuda con la manipulación y análisis de datos estructurados.
    * Carga el dataset (CSV), gestiona el DataFrame y permite filtrar o limpiar registros.
* **`numpy`**
    * Realiza las operaciones matemáticas y manejo de arrays eficientes.
    * Soporte numérico fundamental para las transformaciones vectoriales de los textos.

#### 2. Visualización y Análisis Exploratorio

* **`matplotlib.pyplot`**
    * Generación de gráficos estáticos.
    * Visualización básica de la distribución de clases (Positivo vs. Negativo).
* **`seaborn`**
    * Visualización de datos estadísticos avanzada.
    * Generación de matrices de confusión y gráficos de distribución estéticos para la presentación.
* **`plotly.express`**
    * Permite la creación de gráficos interactivos y dinámicos.
    * Utilizado para generar gráficos de pastel (pie charts) que permiten explorar la distribución de sentimientos de forma interactiva.

#### 3. Procesamiento de Lenguaje Natural (NLP) y Limpieza

* **`re`** (Regular Expressions)
    * Manejo de expresiones regulares.
    * Eliminación de ruido en el texto: URLs, menciones (@usuario), hashtags (#) y caracteres especiales no alfanuméricos.
* **`string`**
    * Constantes de cadenas comunes.
    * Provee listas estándar de signos de puntuación para su eliminación eficiente.
* **`unicodedata`**
    * Facilita la normalización de caracteres Unicode.
    * Crucial para eliminar tildes y diacríticos manteniendo la integridad de letras como la "ñ".
* **`nltk`**
    * Toolkit esencial para el procesamiento de lenguaje natural.
    * Proporciona recursos léxicos y herramientas para el filtrado de palabras irrelevantes (stopwords).

#### 4. Modelado y Machine Learning (Core)

* **`scikit-learn`**
    * Biblioteca principal de Machine Learning.
    * **`TfidfVectorizer`**: Transforma el texto limpio en vectores numéricos.
    * **`LogisticRegression`**: Algoritmo de clasificación supervisada.
    * **`LinearSVC`**: Implementación de SVM para clasificación lineal, optimizada para grandes volúmenes de texto.
    * **`CalibratedClassifierCV`**: Calibra las predicciones para obtener probabilidades de confianza (0-100%).
    * **`GridSearchCV`**: Optimización automática de hiperparámetros mediante búsqueda en malla.
    * **`metrics`**: Cálculo de precisión, recall y F1-score.
    * **`Pipeline`**: Encapsulamiento de los pasos de transformación y predicción.
* **`imblearn (SMOTE)`**
    * Técnica de sobremuestreo sintético para balancear las clases del dataset.
    * Ayuda a que el modelo no se sesgue hacia la clase más frecuente, mejorando la detección de minorías.

#### 5. Persistencia e Integración
Herramientas para conectar el modelo con el Backend.

* **`joblib`**
    * Serialización eficiente de objetos Python.
    * Exportar (`dump`) el pipeline entrenado a un archivo `.joblib` y cargarlo (`load`) en la API para realizar predicciones.
* **`chardet`**
    * Detección automática de la codificación de archivos (encoding).
* **`urllib.request`** & **`json`**
    * Utilidades para realizar peticiones HTTP y procesar datos en formato JSON desde fuentes externas como GitHub.
* **`pathlib`** & **`os`**
    * Gestión de rutas de archivos y directorios de forma independiente al sistema operativo



---



### <font size=16  color=lightgreen> Importando librerías <font>



In [30]:
# Procesamiento y manipulación
import pandas as pd
import numpy as np
# Visualización y analisis exploratorio
import matplotlib.pyplot as plt
import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots
from _plotly_utils.basevalidators import SubplotidValidator
from pathlib import Path
import urllib.response
import urllib.request
from datetime import datetime
import re
import string
import chardet
import unicodedata
from io import StringIO
import uvicorn
import sklearn
import fastapi
import joblib
import nltk
import os
import warnings
import json
from urllib.error import URLError, HTTPError
warnings.filterwarnings('ignore')


In [31]:

# 1. Crear contador global al principio
CONTADOR_GLOBAL = {
    'contradicciones': 0,
    'duplicados': 0,
    'vacios_nulos': 0,
    'sentimientos_nan': 0,
    'total_eliminados' : 0
}


## <font size=12 color=#00FFFF> Extracción, Transformación y Limpieza</font>

### <font size = 8 color="lightgreen">Importación de diccionario<font>

In [32]:
# ==========================================================
# 🎯 CARGAR DICIONARIO DE SENTIMIENTOS (VERSIÓN CORREGIDA)
# ==========================================================

import json
import urllib.request
from urllib.error import URLError, HTTPError

def cargar_diccionario_completo():
    """
    Carga el diccionario de sentimientos y retorna TODAS las variables necesarias.
    """
    # URL corregida (raw de GitHub)
    url = "https://raw.githubusercontent.com/ml-punto-tech/sentiment-api/feature/data-science-marely/data-science/sources/diccionarios/sentimientos_mapeo.json"
    
    print(f"📥 Cargando diccionario desde: {url}")
    
    try:
        # 1. Descargar
        with urllib.request.urlopen(url) as response:
            content = response.read()
        
        # 2. Decodificar y cargar JSON
        datos = json.loads(content.decode('utf-8'))
        print(f"✅ JSON cargado: {len(datos)} sentimientos")
        
        # 3. Crear listas de categorías
        positivos_es = [k for k, v in datos.items() if v == 'positivo']
        negativos_es = [k for k, v in datos.items() if v == 'negativo']
        neutros_es = [k for k, v in datos.items() if v == 'neutral']

        print(f"📊 Distribución:")
        print(f"   • Positivos: {len(positivos_es)}")
        print(f"   • Negativos: {len(negativos_es)}")
        print(f"   • Neutros: {len(neutros_es)}")
        
        # 4. Retornar todas las variables en un diccionario
        return {
            'datos': datos,
            'positivos': positivos_es,
            'negativos': negativos_es,
            'neutros': neutros_es
        }
        
    except HTTPError as e:
        print(f"❌ Error HTTP {e.code}: {e.reason}")
    except URLError as e:
        print(f"❌ Error de URL: {e.reason}")
    except json.JSONDecodeError as e:
        print(f"❌ Error en JSON: {e}")
    except Exception as e:
        print(f"❌ Error inesperado: {type(e).__name__}: {e}")

   # Retornar diccionario vacío en caso de error
    return {
        'datos': {},
        'positivos': [],
        'negativos': [],
        'neutros': []
    }

# ==================== EJECUCIÓN ====================
print("=" * 60)
print("CARGANDO DICCIONARIO DE SENTIMIENTOS")
print("=" * 60)

# Cargar y obtener todas las variables
diccionario_data = cargar_diccionario_completo()

# Extraer las variables individuales
datos_es = diccionario_data['datos']
positivos_es = diccionario_data['positivos']
negativos_es = diccionario_data['negativos']
neutros_es = diccionario_data['neutros']

print("\n" + "=" * 60)
print("✅ VARIABLES CREADAS Y DISPONIBLES:")
print("=" * 60)
print(f"• 'datos_es' (diccionario completo): {len(datos_es)} elementos")
print(f"• 'positivos_es' (lista): {len(positivos_es)} elementos")
print(f"• 'negativos_es' (lista): {len(negativos_es)} elementos")
print(f"• 'neutros_es' (lista): {len(neutros_es)} elementos")

# Mostrar ejemplos
if positivos_es:
    print(f"\n🔍 Ejemplo de positivos: {positivos_es[:3]}")
if negativos_es:
    print(f"🔍 Ejemplo de negativos: {negativos_es[:3]}")
if neutros_es:
    print(f"🔍 Ejemplo de neutros: {neutros_es[:3]}")


CARGANDO DICCIONARIO DE SENTIMIENTOS
📥 Cargando diccionario desde: https://raw.githubusercontent.com/ml-punto-tech/sentiment-api/feature/data-science-marely/data-science/sources/diccionarios/sentimientos_mapeo.json
✅ JSON cargado: 104 sentimientos
📊 Distribución:
   • Positivos: 60
   • Negativos: 39
   • Neutros: 5

✅ VARIABLES CREADAS Y DISPONIBLES:
• 'datos_es' (diccionario completo): 104 elementos
• 'positivos_es' (lista): 60 elementos
• 'negativos_es' (lista): 39 elementos
• 'neutros_es' (lista): 5 elementos

🔍 Ejemplo de positivos: ['aceptacion', 'admiracion', 'adoracion']
🔍 Ejemplo de negativos: ['abrumado', 'aburrimiento', 'aislamiento']
🔍 Ejemplo de neutros: ['ambivalencia', 'anticipacion', 'curiosidad']


In [33]:
#============================================================================
#FUNCIÓN AUXILIAR - PROCESAR DICCIONARIO Y NOMBRES DE VARIABLES
# ============================================================================
def procesar_dic(dict, funcion_proceso, sufijo=''):
    """
    Procesa todos los elementos de un diccionario según su idioma
    Función auxiliar para iteración de diccionarios y creación de nombres actualizados.
    Args:
        dict: Diccionario {nombre_df: dataframe}
        funcion_proceso: Función que procesa un dataframe
        sufijo: Sufijo para el nuevo nombre

    Returns:
        Nuevo diccionario con nombres actualizados
    """
    nuevo_dict = {}

    for nombre, item in dict.items():
        # Extraer partes del nombre
        partes = nombre.split('_')

        if len(partes) >= 2:
            nombre_base = partes[0]          # 'df1'

            # Aplicar función de procesamiento
            df_proc = funcion_proceso(item, nombre)

            # Crear nuevo nombre
            nuevo_nombre = f"{nombre_base}{sufijo}"
            nuevo_dict[nuevo_nombre] = df_proc

            print(f"✅ {nombre} → {nuevo_nombre}")
            print('-' * 80)

    return nuevo_dict

### <font size = 8 color="lightgreen">Importación de datasets<font>

#### **Url Datasets**

In [34]:
datasets = {
    "df1_es":"https://github.com/ml-punto-tech/sentiment-api/raw/refs/heads/dev/data-science/datasets/datasets-origin/dataset1_esp.csv,sep=;",
    "df2_es":"https://github.com/ml-punto-tech/sentiment-api/raw/refs/heads/dev/data-science/datasets/datasets-origin/dataset2_esp.csv,sep=;",
    "df3_es":"https://github.com/ml-punto-tech/sentiment-api/raw/refs/heads/dev/data-science/datasets/datasets-origin/dataset3_esp.csv,sep=,",
}

#### **Función importación dataset**

In [35]:
# IMPORTAR DATASETS

def importar_dataset(url_param, nombre):
    """
    Importa dataset desde URL que incluye parámetros en el string.
    Formato: "url,sep=X" o "url,encoding=Y,sep=Z"
    """
    try:
        print(f"\n{'='*50}")
        print(f"📥 PROCESANDO: {nombre}")
        print(f"{'='*50}")
        
        # 1. Parsear URL y parámetros usando regex para mayor robustez
        if ',' in url_param:
            # Encontrar la URL (todo antes del primer parámetro)
            match = re.match(r'^([^,]+)(,.+)?$', url_param)
            if match:
                url = match.group(1).strip()
                parametros_str = match.group(2) or ""
            else:
                url = url_param
                parametros_str = ""
        else:
            url = url_param
            parametros_str = ""
        
        print(f"🔗 URL: {url[:80]}..." if len(url) > 80 else f"🔗 URL: {url}")
        
        # 2. Extraer parámetros con valores por defecto
        sep = ';'  # separador por defecto
        encoding_param = None
        
        if parametros_str:
            # Extraer todos los parámetros tipo "key=value"
            parametros = re.findall(r'(\w+)=([^,]+)', parametros_str)
            parametros_dict = dict(parametros)
            
            # Obtener separador
            if 'sep' in parametros_dict:
                sep = parametros_dict['sep']
                # Si sep es literal 'comma', convertirlo a ','
                if sep.lower() == 'comma':
                    sep = ','
            
            # Obtener encoding si está especificado
            if 'encoding' in parametros_dict:
                encoding_param = parametros_dict['encoding']
        
        print(f"📝 Separador detectado: '{sep}'")
        if encoding_param:
            print(f"📝 Encoding especificado: {encoding_param}")
        
        # 3. Descargar contenido
        print("⏬ Descargando contenido...")
        req = urllib.request.Request(
            url, 
            headers={'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64)'}
        )
        
        with urllib.request.urlopen(req, timeout=30) as response:
            content = response.read()
        
        # 4. Determinar encoding
        if encoding_param and encoding_param.lower() != 'auto':
            # Usar encoding especificado
            encoding = encoding_param
            print(f"🔍 Usando encoding especificado: {encoding}")
        else:
            # Detectar encoding automáticamente
            print("🔍 Detectando encoding automáticamente...")
            result = chardet.detect(content)
            encoding = result['encoding'] or 'utf-8'
            print(f"🔍 Encoding detectado: {encoding} (confianza: {result['confidence']:.2%})")
        
        # 5. Decodificar y cargar
        print("💾 Cargando DataFrame...")
        decoded_content = content.decode(encoding, errors='replace')
        
        # Intentar cargar con el separador detectado
        try:
            data = pd.read_csv(StringIO(decoded_content), sep=sep)
            print(f"✅ {nombre}: Cargado exitosamente")
            
        except pd.errors.ParserError:
            print(f"⚠️  Error con separador '{sep}'. Probando alternativas...")
            # Probar separadores alternativos comunes
            separadores_alternativos = [',', ';', '\t', '|']
            
            for alt_sep in separadores_alternativos:
                if alt_sep != sep:  # No probar el mismo
                    try:
                        data = pd.read_csv(StringIO(decoded_content), sep=alt_sep)
                        print(f"✅ {nombre}: Cargado con separador alternativo '{alt_sep}'")
                        break
                    except:
                        continue
            else:
                # Si ningún separador funciona, intentar sin especificar
                print("🔄 Intentando con separador automático...")
                data = pd.read_csv(StringIO(decoded_content), sep=None, engine='python')
                print(f"✅ {nombre}: Cargado con separador automático")
        
        # 6. Mostrar información del dataset
        print(f"📊 Dimensiones: {data.shape[0]} filas × {data.shape[1]} columnas")
        print(f"📋 Columnas ({len(data.columns)}):")
        for i, col in enumerate(data.columns[:5]):  # Mostrar primeras 5 columnas
            print(f"   {i+1}. {col}")
        if len(data.columns) > 5:
            print(f"   ... y {len(data.columns)-5} columnas más")
        
        print(f"🔍 Muestra (3 filas aleatorias):")
        print(data.sample(min(3, len(data))).to_string(index=False))
        
        return data
        
    except urllib.error.URLError as e:
        print(f"❌ Error de conexión en {nombre}: {e}")
        return None
    except Exception as e:
        print(f"❌ Error inesperado en {nombre}: {type(e).__name__}: {str(e)[:100]}...")
        return None

# Tu código original SIN MODIFICAR
def fase_carga_datasets(datasets):
    print('=' * 60)
    print('>>> INFORME DE CARGA DATASETS')
    print('=' * 60)
    
    nuevo_diccionario = procesar_dic(
        dict=datasets,
        funcion_proceso=importar_dataset,
        sufijo='_cargado'
    )
    
    # Resumen final
    print('\n' + '=' * 60)
    print('>>> RESUMEN DE CARGA')
    print('=' * 60)
    
    cargados = sum(1 for df in nuevo_diccionario.values() if df is not None)
    print(f"📊 Datasets cargados exitosamente: {cargados}/{len(datasets)}")
    
    for nombre_original, df in zip(datasets.keys(), nuevo_diccionario.values()):
        nombre_nuevo = nombre_original.split('_')[0] + '_cargado'
        if df is not None:
            print(f"  ✅ {nombre_original} → {nombre_nuevo}: {df.shape}")
        else:
            print(f"  ❌ {nombre_original} → {nombre_nuevo}: FALLÓ")
    
    return nuevo_diccionario


# Ejecutar
print("🚀 Iniciando carga de datasets...")
dfs_originales = fase_carga_datasets(datasets)

# Verificar resultados
if dfs_originales:
    print("\n🎯 DATASETS CARGADOS DISPONIBLES:")
    for nombre, df in dfs_originales.items():
        if df is not None:
            print(f"  • {nombre}: {type(df).__name__} con forma {df.shape}")

🚀 Iniciando carga de datasets...
>>> INFORME DE CARGA DATASETS

📥 PROCESANDO: df1_es
🔗 URL: https://github.com/ml-punto-tech/sentiment-api/raw/refs/heads/dev/data-science/d...
📝 Separador detectado: ';'
⏬ Descargando contenido...
🔍 Detectando encoding automáticamente...
🔍 Encoding detectado: utf-8 (confianza: 99.00%)
💾 Cargando DataFrame...
✅ df1_es: Cargado exitosamente
📊 Dimensiones: 1465 filas × 15 columnas
📋 Columnas (15):
   1. Unnamed: 0.1
   2. Unnamed: 0
   3. Text
   4. Sentiment
   5. Timestamp
   ... y 10 columnas más
🔍 Muestra (3 filas aleatorias):
Unnamed: 0.1 Unnamed: 0                                                                                                                    Text Sentiment        Timestamp                User  Platform                        Hashtags Retweets Likes     Country Year Month Day Hour
         256        260 Volando sobre las alas de un espíritu libre, libre de las cadenas del conformismo, pintando el cielo con independencia.   Neutral

In [36]:
# Mostrar las claves del diccionario de dataframes cargados
print(list(dfs_originales.keys()))
df1_cargado = dfs_originales.get('df1_cargado')
df2_cargado = dfs_originales.get('df2_cargado')
df3_cargado = dfs_originales.get('df3_cargado')


TOTAL_INICIAL_REAL = len(df1_cargado) + len(df2_cargado) + len(df3_cargado)
print(f"📊 TOTAL INICIAL REAL (suma de todos los datasets): {TOTAL_INICIAL_REAL:,}")

# Guardar como variable global
INICIAL_GLOBAL = TOTAL_INICIAL_REAL

['df1_cargado', 'df2_cargado', 'df3_cargado']
📊 TOTAL INICIAL REAL (suma de todos los datasets): 4,745


<font color='lightgreen' size=12>Filtrar y explorar datasets</font>

In [37]:
def normalizar_columnas(df):
    df_copia = df.copy()

    col_texto_encontrada = None
    # Buscar una columna que pueda contener el texto
    for nombre_columna in ['texto', 'Text', 'contenido']:
        if nombre_columna in df_copia.columns:
            col_texto_encontrada = nombre_columna
            break

    col_sentimiento_encontrada = None
    # Buscar una columna que pueda contener el sentimiento
    for nombre_columna in ['sentimiento', 'Sentiment', 'etiqueta']:
        if nombre_columna in df_copia.columns:
            # Si la columna 'sentimiento' ya existe y es de tipo string (ej., 'positivo', 'negativo'), priorizarla.
            if nombre_columna == 'sentimiento' and df_copia[nombre_columna].dtype == 'object':
                col_sentimiento_encontrada = nombre_columna
                break
            # De lo contrario, si 'Sentiment' existe y es de tipo objeto
            elif nombre_columna == 'Sentiment' and df_copia[nombre_columna].dtype == 'object':
                col_sentimiento_encontrada = nombre_columna
                break
            # Si 'etiqueta' existe (a menudo numérica para sentimiento)
            elif nombre_columna == 'etiqueta' and pd.api.types.is_numeric_dtype(df_copia[nombre_columna]):
                col_sentimiento_encontrada = nombre_columna
                # No salir inmediatamente, ya que una columna 'sentimiento' de tipo string podría existir y ser preferible.
                # Continuar buscando 'sentimiento' o 'Sentiment' primero.
            # Caso general para cualquier otra columna de sentimiento encontrada
            elif col_sentimiento_encontrada is None: # Solo asignar si no se ha asignado ya una de mayor prioridad.
                col_sentimiento_encontrada = nombre_columna

    if col_texto_encontrada is None or col_sentimiento_encontrada is None:
        # Si las columnas esenciales faltan, retornar un DataFrame vacío con las columnas esperadas.
        return pd.DataFrame(columns=['texto', 'sentimiento'])

    # Renombrar si es necesario
    if col_texto_encontrada != 'texto':
        df_copia.rename(columns={col_texto_encontrada: 'texto'}, inplace=True)
    if col_sentimiento_encontrada != 'sentimiento':
        df_copia.rename(columns={col_sentimiento_encontrada: 'sentimiento'}, inplace=True)

    # Retornar solo las dos columnas requeridas.
    return df_copia[['texto', 'sentimiento']]
df1_filtrado = normalizar_columnas(df1_cargado)
df2_filtrado = normalizar_columnas(df2_cargado)
df3_filtrado = normalizar_columnas(df3_cargado)

In [38]:
def filtrar_dataframe(df,nombre):
    # Añadir una verificación al inicio para DataFrame nulo
    if df is None:
        print(f"Advertencia: DataFrame '{nombre}' es None, saltando filtrado.")
        return None

    # Paso 1: Normalizar nombres de columnas
    # normalizar_columnas ahora devuelve solo 'texto' y 'sentimiento' o un df vacío.
    df = normalizar_columnas(df)

    # Paso 2: Verificar que las columnas necesarias existan después de la normalización
    columnas_requeridas = ['texto', 'sentimiento']
    if not all(col in df.columns for col in columnas_requeridas) or df.empty:
        # Si la normalización falló o devolvió un df vacío, df estará vacío aquí.
        print(f"Advertencia: No se encontraron las columnas requeridas o el DataFrame está vacío después de normalizar en {nombre}.")
        return None

    # Paso 3: La lógica actual de filtrado (ahora se garantiza que df tiene 'texto', 'sentimiento')
    df_filtrado = df[columnas_requeridas].copy()

    # Mostrar estadísticas
    print(f'\nRESUMEN {nombre}')
    print(f"📊 Tamaño del dataframe: {df_filtrado.shape}")
    print(f"📊 Registros únicos: {df_filtrado['texto'].nunique()}")
    print(f"📊 Sentimientos únicos: {df_filtrado['sentimiento'].nunique()}")
    print(f"📊 Textos vacíos: {df_filtrado['texto'].isnull().sum()}")
    print(f"📊 Sentimientos vacíos: {df_filtrado['sentimiento'].isnull().sum()}")
    print(f"📊 Registros duplicados: {df_filtrado.duplicated().sum()}")
    print(f"📊 Textos duplicados: {df_filtrado.duplicated(subset=['texto']).sum()}")

    print(df_filtrado.sample(3))

    return df_filtrado

In [39]:
def fase_filtrado(dfs_originarios):
    # Solo orquesta: delega el recorrido a procesar_dict
    dfs_filtrados = procesar_dic(
        dict=dfs_originales,
        funcion_proceso=filtrar_dataframe,
        sufijo='_filtrado'
    )
    return dfs_filtrados

# Uso
dfs_filtrados = fase_filtrado(dfs_originales)


RESUMEN df1_cargado
📊 Tamaño del dataframe: (1465, 2)
📊 Registros únicos: 708
📊 Sentimientos únicos: 105
📊 Textos vacíos: 2
📊 Sentimientos vacíos: 2
📊 Registros duplicados: 754
📊 Textos duplicados: 756
                                                  texto     sentimiento
1337  Saborear los sabores de una comida casera.Las ...  Contentamiento
504   Al caminar por la Gran Muralla China, cada pas...        Positivo
169   Llega el aburrimiento, el día se siente infini...    Aburrimiento
✅ df1_cargado → df1_filtrado
--------------------------------------------------------------------------------

RESUMEN df2_cargado
📊 Tamaño del dataframe: (2540, 2)
📊 Registros únicos: 2156
📊 Sentimientos únicos: 3
📊 Textos vacíos: 0
📊 Sentimientos vacíos: 0
📊 Registros duplicados: 298
📊 Textos duplicados: 384
                                                  texto sentimiento
1174  Al menda le va a tocar empollar fuerte el Java...    positivo
743   ÁNGEL GONZÁLEZ . Aquí o allí Quién es el que e...    ne

In [40]:
print(list(dfs_filtrados.keys()))

['df1_filtrado', 'df2_filtrado', 'df3_filtrado']


### <font size=12 color=lightgreen>Limpieza y normalización de textos</font>

#### **Función para limpieza de textos**

In [41]:
def limpiar_texto_sentimientos(texto):
    """
    Normaliza texto español preservando ñ y eliminando tildes.
    NO convierte a minúsculas para preservar intensidad emocional.
    """
    # Verifica si la entrada no es una cadena. Si no lo es, devuelve una cadena vacía.
    if not isinstance(texto, str):
        return ""

    # 1. Normaliza el texto para separar los caracteres base de sus diacríticos (ej., tildes).
    texto = unicodedata.normalize('NFD', texto)

    # 2. Reemplaza temporalmente las 'ñ' y 'Ñ' con marcadores especiales para preservarlas
    # durante la eliminación de diacríticos.
    texto = texto.replace('n\u0303', '@@@N_TILDE@@@')
    texto = texto.replace('ñ', '@@@N_TILDE@@@')
    texto = texto.replace('N\u0303', '@@@N_TILDE_MAYUS@@@')
    texto = texto.replace('Ñ', '@@@N_TILDE_MAYUS@@@')

    # 3. Elimina los caracteres diacríticos (como las tildes) del texto.
    texto = ''.join(
        char for char in texto
        if not unicodedata.combining(char)
    )

    # Restaura las 'ñ' y 'Ñ' utilizando los marcadores temporales.
    texto = texto.replace('@@@N_TILDE@@@', 'ñ')
    texto = texto.replace('@@@N_TILDE_MAYUS@@@', 'Ñ')


    # Variable para almacenar el resultado de la limpieza.
    resultado = texto
    chars = []

    # Itera sobre cada caracter en el resultado y añade solo los caracteres imprimibles a una lista.
    # Los caracteres no imprimibles (como los de control) son reemplazados por un espacio.
    for char in resultado:
        if char.isprintable():
            chars.append(char)
        else:
            chars.append(' ')
    resultado = ''.join(chars)
    
    # Elimina los caracteres '#' que están directamente seguidos por una palabra (hashtags).
    resultado = re.sub(r'#(?=\S)', '', resultado)

    # Elimina URLs que terminan en "..." (posibles URLs rotas).
    resultado = re.sub(r'https?://[^\s]*\.\.\.', '[URL_ROTA]', resultado)
    resultado = re.sub(r'www\.[^\s]*\\.\\.\\.', '[URL_ROTA]', resultado)

    # Normaliza los espacios múltiples a uno solo y elimina espacios al inicio y final.
    resultado = ' '.join(resultado.split())
    resultado = resultado.strip()


    # Devuelve el texto preprocesado.
    return resultado


In [42]:
from typing import Iterable
def obtener_lista_ordenada(*series_o_listas: Iterable,nombre='nombre'):
    """
    Acepta cualquier cantidad de Series/Listas de sentimientos y devuelve
    una lista ordenada de sentimientos únicos ya limpios.
    """

    # 1) Unir todas las entradas en un solo iterable
    todos = []
    for s in series_o_listas:
        todos.extend(list(s))


    # 2) Limpiar y eliminar duplicados en un solo paso usando un set
    sentimientos_limpios = {limpiar_texto_sentimientos(x) for x in todos}

    print('\n====> RESUMEN LIMPIEZA',nombre)
    print(f'📊 Registros (original):',len(todos))
    print(f'📊 Registros (despues de la limpieza):',len(sentimientos_limpios))
    # listar 
    print(f'Lista {nombre} limpios: {', '.join(sentimientos_limpios)}')

    print('-' * 80)

    # 3) Devolver ordenado
    return sorted(sentimientos_limpios)

In [43]:
# Limpieza dataframe

def limpiar_columnas(df, col1_name, col2_name, nombre_df):
    df_copy = df.copy()
    
    print(f"🔄 Procesando {nombre_df}...")
    print(f"  Antes: {df_copy.shape}")
    
    df_copy[col1_name + "_limpio"] = df_copy[col1_name].apply(limpiar_texto_sentimientos)
    df_copy[col2_name + "_limpio"] = df_copy[col2_name].apply(limpiar_texto_sentimientos)
    
    print(f"  Después: {df_copy.shape}, nuevas cols: {col1_name}_limpio, {col2_name}_limpio")
    
    return df_copy


In [44]:
# ============================================================================
# FUNCIÓN 3: Fase de limpieza básica (ORQUESTADOR)
# ============================================================================
def fase_limpieza_texto(dfs_filtrados):
    """
    Orquesta la limpieza básica de todos los datasets

    Args:
        dfs_originales: Diccionario con datasets filtrados

    Returns:
        Diccionario con datasets después de limpieza básica
    """
    print("\n" + "="*70)
    print("🚀 FASE 1: LIMPIEZA BÁSICA (COMÚN A TODOS)")
    print("="*70)

    # Usar procesar_dict_con_idioma para recorrer todos
    dfs_limpios_resultado = procesar_dic(
        dict=dfs_filtrados, # Corrected: should use dfs_filtrados here
        funcion_proceso=lambda df, nombre: limpiar_columnas(
            df,
            col1_name="texto",
            col2_name="sentimiento",
            nombre_df=nombre
        ),
        sufijo='_limpio'
    )

    print("\n✅ LIMPIEZA BÁSICA COMPLETADA")
    print(f"• Entrada: {len(dfs_filtrados)} datasets")
    print(f"• Salida: {len(dfs_limpios_resultado)} datasets procesados")

    return dfs_limpios_resultado

# Call the function and assign its result to the global dfs_limpios
dfs_limpios = fase_limpieza_texto(dfs_filtrados)
print(list(dfs_limpios.keys()))


🚀 FASE 1: LIMPIEZA BÁSICA (COMÚN A TODOS)
🔄 Procesando df1_filtrado...
  Antes: (1465, 2)
  Después: (1465, 4), nuevas cols: texto_limpio, sentimiento_limpio
✅ df1_filtrado → df1_limpio
--------------------------------------------------------------------------------
🔄 Procesando df2_filtrado...
  Antes: (2540, 2)
  Después: (2540, 4), nuevas cols: texto_limpio, sentimiento_limpio
✅ df2_filtrado → df2_limpio
--------------------------------------------------------------------------------
🔄 Procesando df3_filtrado...
  Antes: (740, 2)
  Después: (740, 4), nuevas cols: texto_limpio, sentimiento_limpio
✅ df3_filtrado → df3_limpio
--------------------------------------------------------------------------------

✅ LIMPIEZA BÁSICA COMPLETADA
• Entrada: 3 datasets
• Salida: 3 datasets procesados
['df1_limpio', 'df2_limpio', 'df3_limpio']


<font color=lightgreen size=12>Unificación dfs</font>

In [45]:
import pandas as pd

def unificar_dfs_limpios(dfs_limpios):
    """
    Une todos los DataFrames del diccionario dfs_limpios en df_unificado.
    Valida columnas iguales y muestra estadísticas detalladas.
    """
    # Validaciones
    if not dfs_limpios:
        print("❌ Error: dfs_limpios está vacío")
        return None
    
    dfs_list = list(dfs_limpios.values())
    columnas_esperadas = ['texto_limpio', 'sentimiento_limpio']  # De tu workflow
    
    # Estadísticas antes de unir
    print("📊 ESTADÍSTICAS ANTES DE UNIFICAR:")
    print("=" * 50)
    total_filas = 0
    for nombre, df in dfs_limpios.items():
        print(f"{nombre}: {df.shape[0]} filas, {df.shape[1]} cols")
        print(f"  Columnas: {list(df.columns)}")
        if not all(col in df.columns for col in columnas_esperadas):
            print(f"  ⚠️  {nombre} falta columnas esperadas")
        total_filas += df.shape[0]
    
    # Verificar columnas iguales
    columnas_set = {frozenset(df.columns) for df in dfs_list}
    if len(columnas_set) > 1:
        print("❌ Error: Columnas no coinciden entre DataFrames")
        return None
    
    # Unificar
    df_unificado = pd.concat(dfs_list, ignore_index=True)
   
    # Estadísticas después
    print("\n✅ UNIFICACIÓN COMPLETADA:")
    print("=" * 50)
    print(f"df_unificado: {df_unificado.shape[0]} filas, {df_unificado.shape[1]} cols")
    print(f"Total filas originales: {total_filas} ✓")
    print("\nPrimeras 3 filas:")
    print(df_unificado.head(3))
    
    return df_unificado

# Uso:
df_unificado = unificar_dfs_limpios(dfs_limpios)

📊 ESTADÍSTICAS ANTES DE UNIFICAR:
df1_limpio: 1465 filas, 4 cols
  Columnas: ['texto', 'sentimiento', 'texto_limpio', 'sentimiento_limpio']
df2_limpio: 2540 filas, 4 cols
  Columnas: ['texto', 'sentimiento', 'texto_limpio', 'sentimiento_limpio']
df3_limpio: 740 filas, 4 cols
  Columnas: ['texto', 'sentimiento', 'texto_limpio', 'sentimiento_limpio']

✅ UNIFICACIÓN COMPLETADA:
df_unificado: 4745 filas, 4 cols
Total filas originales: 4745 ✓

Primeras 3 filas:
                                              texto sentimiento  \
0      ¡Disfrutando de un hermoso día en el parque!    Positivo   
1              Esta mañana el tráfico era terrible.    Negativo   
2  ¡Acabo de terminar un entrenamiento increíble!??    Positivo   

                                       texto_limpio sentimiento_limpio  
0      ¡Disfrutando de un hermoso dia en el parque!           Positivo  
1              Esta mañana el trafico era terrible.           Negativo  
2  ¡Acabo de terminar un entrenamiento increible!??

### <font color=lightgreen size=12>Limpieza df unificado</font>

In [46]:
def limpieza_dataframe_unificado(df, col_texto='texto_limpio', col_sentimiento='sentimiento_limpio', contador=None):
    """
    Limpieza con acumulación de estadísticas
    """

    print("🧹 LIMPIEZA RÁPIDA CON ESTADÍSTICAS Y AUDITORÍA")
    print("=" * 70)

    # Inicializar contador si no se proporciona
    if contador is None:
        contador = {
            'contradicciones': 0,
            'duplicados': 0,
            'vacios_nulos': 0,
            'sentimientos_nan': 0,
            'total_eliminados': 0

        }

    df_original = df.copy()
    registros_iniciales = len(df)

    # AUDITORÍA CONTRADICCIONES (igual que antes)
    print(f"\n📊 INICIAL: {registros_iniciales:,} registros")
    grupos = df.groupby(col_texto)[col_sentimiento].nunique()
    contradictorios = grupos[grupos > 1]
    print(f"🔍 Textos contradictorios únicos: {len(contradictorios):,}")
    print(f"📈 Registros afectados: {contradictorios.sum():,} (todos se eliminarán)")

    # Mostrar top 3 ejemplos reales (igual que antes)
    top_ejemplos = contradictorios.head(3).index.tolist()
    print("   Ejemplos:")
    for texto in top_ejemplos:
        sentimientos = sorted(df[df[col_texto] == texto][col_sentimiento].unique())
        registros = len(df[df[col_texto] == texto])
        print(f"     ❌ '{texto[:50]}...' → {sentimientos} ({registros} regs)")
    if len(contradictorios) > 3:
        print(f"     ... y {len(contradictorios)-3:,} más")

    # 1. ELIMINAR CONTRADICTORIOS (igual cálculo)
    textos_contradictorios = contradictorios.index.tolist()
    df_antes = df.copy()
    df = df[~df[col_texto].isin(textos_contradictorios)]
    eliminados_contradictorios = len(df_antes) - len(df)

    # ACUMULAR ESTADÍSTICA (NUEVO)
    contador['contradicciones'] += eliminados_contradictorios
    print(f"\n✂️  REGISTROS ELIMINADOS CONTRADICTORIOS: {eliminados_contradictorios:,}")

    # 2. DUPLICADOS (igual cálculo)
    df_antes = df.copy()
    duplicados = df.duplicated(subset=[col_texto, col_sentimiento]).sum()
    df = df.drop_duplicates(subset=[col_texto, col_sentimiento])
    eliminados_duplicados = len(df_antes) - len(df)

    # ACUMULAR ESTADÍSTICA (NUEVO)
    contador['duplicados'] += eliminados_duplicados
    print(f"✂️  REGISTROS ELIMINADOS DUPLICADOS:     {duplicados:,}")

    # 3. VACÍOS Y NULOS (igual cálculo)
    df_antes = df.copy()
    vacios_antes = len(df)
    df = df.dropna(subset=[col_texto, col_sentimiento])
    df = df[(df[col_texto].astype(str).str.strip() != '') &
            (df[col_sentimiento].astype(str).str.strip() != '')]
    vacios_eliminados = len(df_antes) - len(df)

    # ACUMULAR ESTADÍSTICA (NUEVO)
    contador['vacios_nulos'] += vacios_eliminados
    print(f"✂️  REGISTROS ELIMINADOS NULOS/VACÍOS:  {vacios_eliminados:,}")

    # ESTADÍSTICAS FINALES
    registros_finales = len(df)
    total_eliminados = registros_iniciales - registros_finales

    # Update total_eliminados in the contador
    contador['total_eliminados'] += total_eliminados

    print('-' * 80)

    print(f"\n📊 FINAL: {registros_finales:,} registros")
    print(f"📊 ELIMINADOS TOTAL: {total_eliminados:,} ({total_eliminados/registros_iniciales*100:.1f}%)")

    # CREAR ESTADÍSTICAS PARA EL GRÁFICO (NUEVO)
    stats = {
        'inicial': registros_iniciales, # Use registros_iniciales from this specific call
        'final': registros_finales,
        'contradicciones_encontradas': contador['contradicciones'],
        'registros_eliminados_contradicciones': contador['contradicciones'],
        'duplicados_exactos_encontrados': contador['duplicados'],
        'registros_eliminados_duplicados': contador['duplicados'],
        'textos_vacios_eliminados': contador['vacios_nulos'],  # Juntamos vacíos y nulos
        'sentimientos_nan_eliminados': 0,  # No manejas sentimientos NaN separados
        'total_eliminados':contador['total_eliminados'], # This will now be correct
        'porcentaje_eliminado': round((total_eliminados / registros_iniciales * 100), 2) if registros_iniciales > 0 else 0
    }

    # GUARDAR EN EL DATAFRAME (NUEVO)
    df.estadisticas_limpieza = stats

    # Mostrar contador acumulado
    print(f"\n📈 CONTADOR ACUMULADO:")
    print(f"   Contradicciones: {contador['contradicciones']}")
    print(f"   Duplicados: {contador['duplicados']}")
    print(f"   Vacíos/Nulos: {contador['vacios_nulos']}")
    print(f"   Total Eliminados: {contador['total_eliminados']}")

    return df, contador, stats



# 2. Llamar la función pasando el contador
df_limpio, CONTADOR_GLOBAL, stats = limpieza_dataframe_unificado(
    df_unificado,
    contador=CONTADOR_GLOBAL
)


# 3. Verificar que las stats están en df_limpio
print(f"\n✅ Estadísticas guardadas en df: {hasattr(df_limpio, 'estadisticas_limpieza')}")

🧹 LIMPIEZA RÁPIDA CON ESTADÍSTICAS Y AUDITORÍA

📊 INICIAL: 4,745 registros
🔍 Textos contradictorios únicos: 90
📈 Registros afectados: 180 (todos se eliminarán)
   Ejemplos:
     ❌ '"De manera apacible, se puede sacudir el mundo" MG...' → ['negativo', 'positivo'] (3 regs)
     ❌ '"He aprendido que el valor no es la ausencia de mi...' → ['neutral', 'positivo'] (2 regs)
     ❌ '"La soledad es peligrosa. Es muy adictiva. Se conv...' → ['negativo', 'positivo'] (6 regs)
     ... y 87 más

✂️  REGISTROS ELIMINADOS CONTRADICTORIOS: 216
✂️  REGISTROS ELIMINADOS DUPLICADOS:     1,073
✂️  REGISTROS ELIMINADOS NULOS/VACÍOS:  1
--------------------------------------------------------------------------------

📊 FINAL: 3,455 registros
📊 ELIMINADOS TOTAL: 1,290 (27.2%)

📈 CONTADOR ACUMULADO:
   Contradicciones: 216
   Duplicados: 1073
   Vacíos/Nulos: 1
   Total Eliminados: 1290

✅ Estadísticas guardadas en df: True


In [47]:
# PRIMERA NORMALIZACION - DF LIMPIO

def primera_normalizacion(df_limpio):
    """
    Versión simple de normalización
    """
    if df_limpio is None or df_limpio.empty:
        print("❌ Dataset vacío o None")
        return None
    
    print("🔧 Normalizando dataset...")
    
    # 1. Eliminar columnas originales si existen
    columnas_a_eliminar = []
    if 'texto' in df_limpio.columns:
        columnas_a_eliminar.append('texto')
    if 'sentimiento' in df_limpio.columns:
        columnas_a_eliminar.append('sentimiento')
    
    if columnas_a_eliminar:
        df = df_limpio.drop(columns=columnas_a_eliminar)
        print(f"   Eliminadas columnas: {columnas_a_eliminar}")
    else:
        df = df_limpio.copy()
    
    # 2. Renombrar columnas limpias
    mapeo = {}
    if 'texto_limpio' in df.columns:
        mapeo['texto_limpio'] = 'texto'
    if 'sentimiento_limpio' in df.columns:
        mapeo['sentimiento_limpio'] = 'sentimiento'
    
    if mapeo:
        df = df.rename(columns=mapeo)
        print(f"   Renombradas columnas: {mapeo}")
    
    # 3. Verificar resultado
    print(f"\n✅ Dataset normalizado:")
    print(f"   • Forma: {df.shape}")
    print(f"   • Columnas: {list(df.columns)}")
    
    # Verificar que tengamos las columnas esenciales
    if 'texto' not in df.columns:
        print("⚠️  Advertencia: Columna 'texto' no encontrada")
    if 'sentimiento' not in df.columns:
        print("⚠️  Advertencia: Columna 'sentimiento' no encontrada")
    
    return df
df_normal1 = primera_normalizacion(df_limpio)

🔧 Normalizando dataset...
   Eliminadas columnas: ['texto', 'sentimiento']
   Renombradas columnas: {'texto_limpio': 'texto', 'sentimiento_limpio': 'sentimiento'}

✅ Dataset normalizado:
   • Forma: (3455, 2)
   • Columnas: ['texto', 'sentimiento']


In [48]:
df_normal1.head()

texto sentimiento
0       ¡Disfrutando de un hermoso dia en el parque!    Positivo
1               Esta mañana el trafico era terrible.    Negativo
2   ¡Acabo de terminar un entrenamiento increible!??    Positivo
3  ¡Emocionado por la escapada de fin de semana q...    Positivo
4   Probando una nueva receta para cenar esta noche.     Neutral

### <font size=12 color=lightgreen>Categorizar de sentimientos </font>

In [49]:
categorias ='positivo_es, negativo_es, neutral_es'
def verificar_sentimientos_clasificados():
    """
    Verifica qué sentimientos en df_unificado no están clasificados en datos_es.
    Muestra estadísticas y lista de sentimientos no clasificados.
    """
    print("\n🔍 VERIFICANDO SENTIMIENTOS NO CLASIFICADOS")
    print("=" * 70)
# Identificar y mostrar entimientos que están en sentimientos_unicos_es, pero no en datos_es.keys
sentimientos_unicos = sorted(df_unificado['sentimiento_limpio'].unique())
print(f'Total de sentimientos: {len(sentimientos_unicos)}')
print('Sentimientos únicos:', sentimientos_unicos)

# Convertir las claves de datos_es a minúsculas para una comparación sin distinción de mayúsculas y minúsculas
datos_es_lower = {k.lower() for k in datos_es.keys()}

sentimientos_no_clasificados = []
for sentimiento in sentimientos_unicos:
    # Limpiar y convertir a minúsculas para la comparación
    sentimiento_limpio_lower = sentimiento.strip().lower()
    # Excluir la cadena vacía si es que existe
    if sentimiento_limpio_lower and sentimiento_limpio_lower not in datos_es_lower:
        sentimientos_no_clasificados.append(sentimiento)

print(f'Sentimientos no clasificados (total: {len(sentimientos_no_clasificados)}): ')

if sentimientos_no_clasificados:
    print(f" Son: {', '.join(sentimientos_no_clasificados)}")
else:
    print("No se encontraron sentimientos no clasificados.")


Total de sentimientos: 109
Sentimientos únicos: ['', 'Abrumado', 'Aburrimiento', 'Aceptacion', 'Admiracion', 'Adoracion', 'Agradecido', 'Aislamiento', 'Alegria', 'Amabilidad', 'Amargura', 'Ambivalencia', 'Amistad', 'Amor', 'Angustia', 'Anhelo', 'Animo', 'Ansiedad', 'Anticipacion', 'Apreciacion', 'Aprensivo', 'Armonia', 'Arrepentimiento', 'Asco', 'Asombro', 'Cautivacion', 'Celebracion', 'Colorido', 'Confiado', 'Confianza', 'Contentamiento', 'Creatividad', 'Cumplimiento', 'Curiosidad', 'Decepcion', 'Desamor', 'Descubrimiento', 'Desesperacion', 'Deslumbrar', 'Despectivo', 'Determinacion', 'Devastado', 'Disfrute', 'Diversion', 'Dolor', 'Elegancia', 'Emocion', 'Empatico', 'Empoderamiento', 'Encantamiento', 'Energia', 'Enojo', 'Entumecimiento', 'Entusiasmo', 'Envidia', 'Envidioso', 'Esperanza', 'Euforia', 'Excitacion', 'Exito', 'Felicidad', 'Frustracion', 'Frustrado', 'Grandeza', 'Gratitud', 'Inspiracion', 'Inspirado', 'Intimidacion', 'Jugueton', 'Lastima', 'Logro', 'Malo', 'Maravilla', 'Mel

#### **Función para categorizar sentimientos**

In [50]:
# SEGUNDA NORMALIZACIÓN
def segunda_normalizacion(df):
	# Crea una copia con un dataset excluyendo valores nulos
	normalizado = df[df['sentimiento'].notna()].copy()
	# Quitar la columna sentimiento
	normalizado = normalizado.drop(columns=['sentimiento']).reset_index(drop=True)
	# Cambiar nombre de la columna sentimiento por sentimiento_final
	normalizado = normalizado.rename(columns={'sentimiento_final': 'sentimiento'})

	return normalizado

In [51]:
def categorizar_sentimiento(sentimiento, categorias, nombres=('positivo', 'negativo', 'neutral')):

    """
    Versión flexible que permite nombres personalizados para las categorías.
    """
    if pd.isna(sentimiento):
        return None
    
    sent = str(sentimiento).strip().lower()
    
    # Iterar sobre cada categoría
    for i, lista_categoria in enumerate(categorias):
        if sent in lista_categoria:
            return nombres[i]

    return None


In [52]:
df_normal1.sample(3)

texto sentimiento
140    Gratitud por la comunidad de apoyo que me rodea.    Gratitud
3587  Dios mio si no me tomo un tamarindo con limon ...     neutral
2533  ¿Va siendo hora de un cafe no? En realidad cua...    negativo

#### **Categorizar sentimientos**

In [53]:
positivos_lista_es = obtener_lista_ordenada(positivos_es, nombre='positivos_es')
negativos_lista_es = obtener_lista_ordenada(negativos_es, nombre='negativos_es')
neutros_lista_es = obtener_lista_ordenada(neutros_es, nombre='neutros_es')
categorias =[positivos_lista_es, negativos_lista_es, neutros_lista_es]
nombres_categorias = ('positivo', 'negativo', 'neutral')

# Aplicar categorización
df_normal1['sentimiento_final'] = df_normal1['sentimiento'].apply(
    lambda x: categorizar_sentimiento(x,categorias, nombres_categorias)
    )
print(df_normal1.sample(3))

# Segunda normalización
df_normalizado = segunda_normalizacion(df_normal1)
print(f"✅ df_normalizado: {len(df_normalizado)} registros categorizados")

df_normalizado.sample(5)


====> RESUMEN LIMPIEZA positivos_es
📊 Registros (original): 60
📊 Registros (despues de la limpieza): 60
Lista positivos_es limpios: positividad, determinacion, inspiracion, gratitud, ternura, inspirado, celebracion, reconfortante, esperanza, asombro, positivo, descubrimiento, animo, apreciacion, amabilidad, elegancia, amistad, euforia, motivacion, amor, aceptacion, exito, creatividad, deslumbrar, excitacion, triunfo, romance, resplandor, serenidad, alegria, grandeza, jugueton, resiliencia, energia, disfrute, empoderamiento, encantamiento, maravilla, adoracion, colorido, confianza, diversion, emocion, optimismo, reverencia, melodico, satisfaccion, confiado, admiracion, logro, agradecido, intimidacion, cautivacion, contentamiento, empatico, entusiasmo, felicidad, cumplimiento, armonia, orgullo
--------------------------------------------------------------------------------

====> RESUMEN LIMPIEZA negativos_es
📊 Registros (original): 39
📊 Registros (despues de la limpieza): 39
Lista nega

texto sentimiento
662   Sentirse solo un sabado por la noche.A veces l...    negativo
2156  sigo sintiendo la misma paz cada vez que hablo...     neutral
328   Un alegre reencuentro con amigos perdidos hace...    positivo
1716  ?Para entender el mundo, lee. ?Para entenderte...    positivo
2597  Si Sura se va, solo perdemos nosotros. Ellos y...    positivo

In [54]:
df_normalizado.sample(5)

texto sentimiento
2036  Vengo a divulgar tu nombre San Cipriano en agr...    positivo
393   Inundado de serenidad mientras el sol se pone ...    positivo
2987  Quizas publique muchas capturas de pantalla qu...     neutral
1185  El Ayuntamiento de Madrid reprueba a Ortega Sm...    negativo
1238  Señore porque dañar a alguien por las redes, s...    negativo

<font color=lightgreen size=12>Informe final registros eliminados y visualización</font>

#### **Función limpieza dataset unificado**

In [55]:
df_normalizado.head(2)

texto sentimiento
0  ¡Disfrutando de un hermoso dia en el parque!    positivo
1          Esta mañana el trafico era terrible.    negativo

In [56]:
# SEGUNDO PROCESO DE LIMPIEZA

# 1.Llamar la función pasando el contador
df_final, CONTADOR_GLOBAL, stats_temp = limpieza_dataframe_unificado(
    df_normalizado, col_texto='texto',col_sentimiento='sentimiento',
    contador=CONTADOR_GLOBAL
)

# Actualizar las estadísticas para el gráfico con los valores globales
stats_final = {
    'inicial': INICIAL_GLOBAL, # Usar la variable global para el total inicial de todos los datasets
    'final': len(df_final),
    'contradicciones_encontradas': stats_temp['contradicciones_encontradas'],
    'registros_eliminados_contradicciones': stats_temp['registros_eliminados_contradicciones'],
    'duplicados_exactos_encontrados': stats_temp['duplicados_exactos_encontrados'],
    'registros_eliminados_duplicados': stats_temp['registros_eliminados_duplicados'],
    'textos_vacios_eliminados': stats_temp['textos_vacios_eliminados'],
    'sentimientos_nan_eliminados': stats_temp['sentimientos_nan_eliminados'],
    'total_eliminados': CONTADOR_GLOBAL['total_eliminados'], # Usar el contador global acumulado
    'porcentaje_eliminado': round((CONTADOR_GLOBAL['total_eliminados'] / INICIAL_GLOBAL * 100), 2) if INICIAL_GLOBAL > 0 else 0
}

df_final.estadisticas_limpieza = stats_final
# 2. Verificar que las stats están en df_limpio
# Esto está BIEN (verifica en df_final):
print(f"\n✅ Estadísticas guardadas en df: {hasattr(df_final, 'estadisticas_limpieza')}")

🧹 LIMPIEZA RÁPIDA CON ESTADÍSTICAS Y AUDITORÍA

📊 INICIAL: 3,455 registros
🔍 Textos contradictorios únicos: 0
📈 Registros afectados: 0 (todos se eliminarán)
   Ejemplos:

✂️  REGISTROS ELIMINADOS CONTRADICTORIOS: 0
✂️  REGISTROS ELIMINADOS DUPLICADOS:     0
✂️  REGISTROS ELIMINADOS NULOS/VACÍOS:  1
--------------------------------------------------------------------------------

📊 FINAL: 3,454 registros
📊 ELIMINADOS TOTAL: 1 (0.0%)

📈 CONTADOR ACUMULADO:
   Contradicciones: 216
   Duplicados: 1073
   Vacíos/Nulos: 2
   Total Eliminados: 1291

✅ Estadísticas guardadas en df: True


In [57]:
import plotly.graph_objects as go
from plotly.subplots import make_subplots

# Obtener las estadísticas de limpieza
if hasattr(df_final, 'estadisticas_limpieza'):
    stats = df_final.estadisticas_limpieza
else:
    print("Error: No se encontraron las estadísticas de limpieza en df_final.")
    # Crear un diccionario de stats de ejemplo para que el código no falle en caso de error
    stats = {
        'inicial': 0,
        'final': 0,
        'contradicciones_encontradas': 5,
        'registros_eliminados_contradicciones': 10,
        'duplicados_exactos_encontrados': 20,
        'registros_eliminados_duplicados': 20,
        'textos_vacios_eliminados': 3,
        'sentimientos_nan_eliminados': 2,
        'total_eliminados': 30,
        'porcentaje_eliminado': 30.0
    }

# --- Datos para el primer subplot (gráfico de barras horizontales) ---
bar_labels_unsorted = ['Contradicciones', 'Duplicados', 'Vacíos/NaN']
bar_values_unsorted = [
    stats['registros_eliminados_contradicciones'],
    stats['registros_eliminados_duplicados'],
    stats['textos_vacios_eliminados'],
    stats['sentimientos_nan_eliminados']
]

# Combinar y ordenar los datos para el gráfico de barras en orden ascendente (al revés del anterior)
sorted_bars = sorted(zip(bar_values_unsorted, bar_labels_unsorted), reverse=False)
bar_values = [val for val, label in sorted_bars]
bar_labels_raw = [label for val, label in sorted_bars] # Keep raw labels for Y-axis

# Calcular porcentajes y crear etiquetas combinadas para el texto en las barras
total_eliminated_for_bars = sum(bar_values) if sum(bar_values) > 0 else 1
bar_text_labels = []
for i, label in enumerate(bar_labels_raw):
    count = bar_values[i]
    percentage = (count / total_eliminated_for_bars) * 100
    bar_text_labels.append(f"{count:,}<br>({percentage:.1f}%)")


# --- Datos para el segundo subplot (gráfico circular) ---
porcentaje_eliminado = stats['porcentaje_eliminado']
porcentaje_sin_eliminar = 100 - porcentaje_eliminado
pie_labels = ['Registros Eliminados', 'Registros Conservados']
pie_values = [
    stats['total_eliminados'], # Usar el total de eliminados
    stats['final'] # Usar el total final de registros
]

# Colores para el pie chart (ej. rojo para eliminados, verde para conservados)
pie_colors = ['#EF553B', '#636EFA'] # Rojo para eliminados, azul para conservados

# Crear subplots
fig = make_subplots(rows=1, cols=2, specs=[[{"type": "xy"}, {"type": "domain"}]]
                    ,subplot_titles=('Estadísticas de Eliminación de Registros', 'Distribución Final de Registros'))

# Añadir gráfico de barras horizontales
fig.add_trace(go.Bar(y=bar_labels_raw, x=bar_values, orientation='h',
    marker_color=['#00CC96', '#9400D3', '#0000CD'],
    showlegend=False,
    text=bar_text_labels,
    textposition='auto'),
    row=1, col=1)

# Añadir gráfico circular
fig.add_trace(go.Pie(labels=pie_labels, values=pie_values, name="Eliminación",
    marker_colors=pie_colors, textinfo='percent+value', hole=.2, insidetextfont=dict(color='white', size=14),
    marker=dict(colors=pie_colors)),
    row=1, col=2)

# Actualizar layout
fig.update_layout(
    title_text=f'<b>Análisis Detallado del Proceso de Limpieza del Dataframe</b><br><sup>Total de Registros Originales: {stats["inicial"]}</sup>',
    title_x=0.5,
    showlegend=True,
    height=500,
    width=1000,
    margin=dict(t=150)
)

fig.show()

 ### <font size=12 color=lightgreen>Análisis de Distribución y Visualización</font>

#### **Análisis de distribución de sentimientos**

In [58]:
#📊 ANÁLISIS DE DISTRIBUCIÓN DEL DATASET FINAL

print("=" * 80)
print("📈 ANÁLISIS DE DISTRIBUCIÓN - DATASET FINAL")
print("=" * 80)

# 1. Calcular conteos y porcentajes
conteos = df_final['sentimiento'].value_counts()
total_registros = len(df_final)
porcentajes = (conteos / total_registros * 100).round(2)

# 2. Mostrar tabla detallada
print(f"{'SENTIMIENTO':<12} | {'CANTIDAD':>8} | {'PORCENTAJE':>10} | {'PROPORCIÓN'}")
print("-" * 80)

for sentimiento in ['positivo', 'negativo', 'neutral']:
    if sentimiento in conteos:
        count = conteos[sentimiento]
        porcentaje = porcentajes[sentimiento]
        # Crear barra visual
        barra = '█' * int(count / total_registros * 40)  # Escala a 40 caracteres
        print(f"{sentimiento.capitalize():<12} | {count:>8} | {porcentaje:>9}% | {barra}")

print("-" * 80)
print(f"{'TOTAL':<12} | {total_registros:>8} | {'100.00':>9}% | {'█' * 40}")
print("-" * 80)

📈 ANÁLISIS DE DISTRIBUCIÓN - DATASET FINAL
SENTIMIENTO  | CANTIDAD | PORCENTAJE | PROPORCIÓN
--------------------------------------------------------------------------------
Positivo     |     1199 |     34.71% | █████████████
Negativo     |     1113 |     32.22% | ████████████
Neutral      |     1142 |     33.06% | █████████████
--------------------------------------------------------------------------------
TOTAL        |     3454 |    100.00% | ████████████████████████████████████████
--------------------------------------------------------------------------------


#### **Visualización de la distribución de Sentimientos**

In [71]:
valores = df_final['sentimiento'].value_counts().reset_index()
valores.columns = ['sentimientos', 'Cantidad']

# Capitalizar las etiquetas para el gráfico circular
labels_capitalized = valores.sentimientos.apply(lambda x: x.capitalize())

# Define custom colors for consistency
sentiment_colors = {
    'positivo': '#EF553B',  # Orange-Red
    'negativo': '#636EFA',   # Blue
    'neutral': '#00CC96'   # Greenish-Teal (a valid color for neutral)
}

pie_colors = [sentiment_colors[s.lower()] for s in labels_capitalized]

bar_colors = [sentiment_colors[s.lower()] for s in valores.sentimientos]

# Crear gráfico subplot con gráfico circular y columnas, especificando los tipos de subplot
fig = make_subplots(rows=1, cols=2, specs=[[{"type": "domain"}, {"type": "xy"}]])

# gráfico circular
fig.add_trace(go.Pie(labels=labels_capitalized, values=valores.Cantidad,
    textposition='inside', textinfo='label+percent',hole=.2,
    insidetextfont=dict(color='white', size=14),
    marker=dict(colors=pie_colors)),
    row=1, col=1)

# gráfico de barras
fig.add_trace(go.Bar(x=valores.sentimientos, y=valores.Cantidad,
    marker=dict(color=bar_colors),
    text=valores.Cantidad, # Add the counts as text labels
    textposition='auto'), # Automatically position the text labels
    row=1, col=2)

# Añadir un título general al subplot
fig.update_layout(
    title_text=f'<b>Distribución de Sentimientos</b><br><span style="font-size:14px">Dataset Final: {total_registros} registros</span>',
    title_x=0.5,
    showlegend=False,
    height=500,
    width=1000
)

fig.show()

### <font size=12 color=lightgreen> Exportar dataset </font>

#### **Definir ruta de exportación**

In [60]:
# Ruta actual
ruta_actual = Path.cwd()

# Buscar data-science
if ruta_actual.name == 'notebooks':
    # Si estamos en notebooks/, ir a ../datasets
    carpeta_datasets = ruta_actual.parent / 'datasets'
else:
    # Buscar data-science en directorios padres
    for directorio_padre in ruta_actual.parents:
        if (directorio_padre / 'data-science').exists():
            carpeta_datasets = directorio_padre / 'data-science' / 'datasets'
            break
    else:
        # Si no encuentra, usar directorio actual/datasets
        carpeta_datasets = ruta_actual / 'datasets'

# Crear carpeta si no existe
carpeta_datasets.mkdir(parents=True, exist_ok=True)

# Ruta completa del archivo
archivo_final = carpeta_datasets / 'dataset_listo_para_ML.csv'


#### **Exportar dataset**

In [61]:
# Renombrar columnas para formato final
df_exportar = df_final.rename({
    'Texto_Limpio': 'texto',
    'Sentimiento_Final': 'sentimiento'
}, axis=1)


metadata = {
    "total_registros": len(df_exportar),
    "distribucion": dict(df_exportar['sentimiento'].value_counts()),
    "fecha_creacion": datetime.now().isoformat(),
    "version": "1.0.0",
    "fuentes": [
        "sentimentdataset_es.csv",
        "sentiment_analysis_dataset.csv"
    ]
}

# Exportar
df_exportar.to_csv(archivo_final, index=False, encoding='utf-8-sig')
print(f"✅ Dataset exportado: {archivo_final}")
print(f"📊 Registros: {len(df_exportar):,}")

# Crear copia para trabajo posterior
df = df_exportar.copy()

✅ Dataset exportado: c:\Users\marely\OneDrive\Documentos\Oracle_ONE\Hackaton\SentimentAPI-Project\sentiment-api\data-science\datasets\dataset_listo_para_ML.csv
📊 Registros: 3,454


#### **Verificar exportación**

In [62]:
def verificar_csv_simple(ruta_archivo, mostrar_muestra=True):
    """
    Verificación simplificada con detección de encoding
    Y verificación de integridad mejorada
    """
    ruta = Path(ruta_archivo)

    if not ruta.exists():
        print(f"❌ Archivo no encontrado: {ruta}")
        return None

    # Detectar encoding
    encodings = ['utf-8-sig', 'utf-8', 'latin1', 'cp1252']

    for enc in encodings:
        try:
            # Probar con 5 filas primero
            df_test = pd.read_csv(ruta, encoding=enc, nrows=5)

            # Si llegamos aquí, el encoding funciona
            try:
                # Ahora cargar completo
                df = pd.read_csv(ruta, encoding=enc)
                print(f"✅ CSV cargado: {len(df):,} registros (encoding: {enc})")

                # 🔍 VERIFICACIÓN DE INTEGRIDAD MEJORADA
                print("🔍 Verificación de integridad:")
                print(f"   • Valores nulos totales: {df.isnull().sum().sum()}")
                print(f"   • Textos vacíos: {(df['texto'].str.strip() == '').sum()}")

                # Verificar que todos los sentimientos sean válidos
                sentimientos_validos = ['positivo', 'negativo', 'neutral']
                sentimientos_invalidos = df[~df['sentimiento'].isin(sentimientos_validos)]

                if len(sentimientos_invalidos) > 0:
                    print(f"   ⚠️  Sentimientos inválidos: {len(sentimientos_invalidos)}")
                    print(f"      Valores únicos inválidos: {sentimientos_invalidos['sentimiento'].unique()}")
                else:
                    print(f"   ✅ Todos los sentimientos son válidos")

                # Verificar unicidad
                textos_unicos = df['texto'].nunique()
                if len(df) == textos_unicos:
                    print(f"   ✅ 100% textos únicos: {textos_unicos:,} textos únicos")
                else:
                    print(f"   ⚠️  Duplicados: {len(df) - textos_unicos:,} textos duplicados")

                if mostrar_muestra:
                    print(f"📝 Columnas: {list(df.columns)}")
                    print(f"📊 Muestra (2 filas):")
                    print(df.head(2).to_string(index=False))

                return df

            except Exception as e:
                print(f"❌ Error cargando con encoding {enc}: {type(e).__name__}")
                continue

        except UnicodeDecodeError:
            continue

    print("❌ No se pudo cargar con ningún encoding común")
    return None

In [63]:
# Uso simple - así debería funcionar
df_check = verificar_csv_simple(archivo_final, mostrar_muestra=True)

✅ CSV cargado: 3,454 registros (encoding: utf-8-sig)
🔍 Verificación de integridad:
   • Valores nulos totales: 0
   • Textos vacíos: 0
   ✅ Todos los sentimientos son válidos
   ✅ 100% textos únicos: 3,454 textos únicos
📝 Columnas: ['texto', 'sentimiento']
📊 Muestra (2 filas):
                                       texto sentimiento
¡Disfrutando de un hermoso dia en el parque!    positivo
        Esta mañana el trafico era terrible.    negativo


 ### <font size=12 color=lightgreen> Resumen ejecutivo </font>

In [64]:
print("=" * 70)
print("📋 RESUMEN EJECUTIVO - HACKATHON SENTIMENT API")
print("=" * 70)
print(f"✅ Dataset final: {len(df_exportar):,} registros")
print(f"✅ Distribución balanceada: {porcentajes['positivo']}% 👍 | {porcentajes['negativo']}% 👎 | {porcentajes['neutral']}% 😐")
print(f"✅ Calidad del dataset:")
print(f"   • 0 contradicciones (cada texto tiene único sentimiento)")
print(f"   • 0 duplicados (100% textos únicos)")
print(f"   • 0 valores nulos")

📋 RESUMEN EJECUTIVO - HACKATHON SENTIMENT API
✅ Dataset final: 3,454 registros
✅ Distribución balanceada: 34.71% 👍 | 32.22% 👎 | 33.06% 😐
✅ Calidad del dataset:
   • 0 contradicciones (cada texto tiene único sentimiento)
   • 0 duplicados (100% textos únicos)
   • 0 valores nulos


---
### <font size=12 color=lightgreen>Observaciones</font>


# 🎯 RESUMEN DE DESAFÍOS ENCONTRADOS Y SOLUCIONES IMPLEMENTADAS

Durante el desarrollo del pipeline de procesamiento, identificamos **6 desafíos principales** que requirieron soluciones específicas y decisiones fundamentadas.

## ORIGEN DE DATOS
Con el objetivo de mejorar la capacidad de generalización del modelo, se trabajó con dos datasets independientes obtenidos desde Kaggle.
Si bien ambos conjuntos de datos abordan el análisis de sentimiento en español, presentan diferencias en estructura, calidad lingüística y formato de origen. Su integración permitió ampliar la diversidad de expresiones textuales, reduciendo el sesgo hacia un único estilo de redacción y fortaleciendo la robustez del pipeline de preparación de datos en escenarios similares a producción.

### (Kaggle):**

- DATASET1_ES ==> https://www.kaggle.com/datasets/engineercolsoquas/spanish-sentiment-analysis-dataset

- DATASET2_ES ==> https://www.kaggle.com/datasets/kashishparmar02/social-media-sentiments-analysis-dataset

- DATASET3_ES ==> https://www.kaggle.com/datasets/jp797498e/twitter-entity-sentiment-analysis

## 🔍 DESAFÍOS IDENTIFICADOS Y SOLUCIONES IMPLEMENTADAS

### 1️⃣ 🌍 **DESAFÍO: DATASETS EN INGLÉS REQUIEREN TRADUCCIÓN A ESPAÑOL**

**Contexto:** Dataset1 y Dataset3 originalmente en inglés  
**Problema:** Modelo final necesita consistencia lingüística en español  
**Riesgo:** Mezcla de idiomas introduce ruido en embeddings y clasificación

**✅ SOLUCIÓN: PROCESO DE TRADUCCIÓN DE DOS FASES**

**Fase 1 - Automatización:**
- APIs de traducción (Google Translate, DeepL)
- Procesamiento batch para escalabilidad

**Fase 2 - Revisión manual:**
- Hablantes nativos corrigen matices emocionales
- Excel para revisión colaborativa
- Corrección de falsos amigos y expresiones idiomáticas

**📊 JUSTIFICACIÓN:**
- **Ejemplo crítico:** `'This is sick!'` → `'¡Esto es increíble!'` (no literal)
- Traducción palabra-por-palabra pierde polaridad emocional
- Inversión en traducción paga en calidad final del dataset

---

### 2️⃣ 🔠 **DESAFÍO: INCONSISTENCIAS DE ENCODING ENTRE DATASETS**

**Contexto:** Cada dataset con encoding diferente (UTF-8, Windows-1252, etc.)  
**Problema:** Caracteres corruptos (�), tildes perdidas, 'ñ' dañada  
**Riesgo:** Pérdida de significado y ruido en procesamiento NLP

**✅ SOLUCIÓN: DETECCIÓN AUTOMÁTICA Y NORMALIZACIÓN UNIFICADA**

- **Herramienta:** `chardet` para detección automática de encoding
- **Proceso:** `normalizar_texto()` con manejo específico de caracteres españoles
- **Preservación:** Mantener 'ñ' y eliminar tildes inteligentemente
- **Validación:** Verificar que `'niño'` → `'niño'` (no `'nino'`)

**📊 JUSTIFICACIÓN:**
- Encoding incorrecto corrompe análisis léxico
- `'cañón'` ≠ `'canon'` (significados completamente diferentes)
- Normalización consistente esencial para modelos basados en tokens

---

### 3️⃣ 🎓 **DESAFÍO: CARACTERÍSTICAS QUE SUGIEREN MATERIAL DE ENTRENAMIENTO**

**Observación:** Patrones repetitivos y estructuras didácticas  
**Dataset2:** Contradicciones intencionales (mismo texto, diferente etiqueta)  
**Dataset3:** Variaciones lingüísticas pedagógicas (6 formas de decir lo mismo)  
**Riesgo:** Dataset no representa distribución real del lenguaje

**✅ SOLUCIÓN: ANÁLISIS Y LIMPIEZA ADAPTATIVA POR PATRÓN**

- **Para contradicciones (Dataset2):** Eliminación completa (216 registros)
- **Para variaciones (Dataset3):** Conservación con documentación
- **Análisis:** Identificar clusters temáticos (ej: 'Borderlands murder')
- **Documentación:** Registrar patrones encontrados para transparencia

**📊 JUSTIFICACIÓN:**
- **Contradicciones:** Mejor eliminar que entrenar con etiquetas incorrectas
- **Variaciones:** Conservar como ejemplos de equivalencia semántica
- **Transparencia:** Documentar hallazgos para usuarios futuros

---

### 4️⃣ 🏷️ **DESAFÍO: GRANULARIDAD FINA EN SENTIMIENTOS (Dataset1)**

**Contexto:** Dataset1 tiene 105 sentimientos específicos  
**Ejemplos:** `'admiración'`, `'asombro'`, `'respeto'`, `'adoración'`  
**Problema:** Demasiadas clases para clasificación efectiva  
**Riesgo:** Overfitting y dificultad en generalización

**✅ SOLUCIÓN: DICCIONARIO DE MAPEO A 3 CATEGORÍAS PRINCIPALES**

- **Fuente:** Diccionario externo con 106 sentimientos mapeados
- **Categorías:** `positivo`, `negativo`, `neutral`
- **Proceso:** `categorizar_sentimiento()` con búsqueda en diccionario
- **Validación:** Verificar mapeos controvertidos manualmente

**📊 JUSTIFICACIÓN:**
- **105 clases → 3 clases:** Reducción dimensional manejable
- **Diccionario externo:** Aprovecha trabajo curado existente
- **Consistencia:** Mismo mapeo para `'admiración'` y `'asombro'` → `'positivo'`

---

### 5️⃣ 💬 **DESAFÍO: DIALECTO DE REDES SOCIALES Y LENGUAJE INFORMAL**

**Contexto:** Textos de Twitter, comentarios, mensajes informales  
**Características:** Abreviaciones, emoticonos, hashtags, lenguaje coloquial  
**Ejemplos:** `'xq'`, `'tb'`, `'q'`, `'???'`, `'lol'`, hashtags emocionales  
**Riesgo:** Procesamiento literal pierde significado emocional

**✅ SOLUCIÓN: LIMPIEZA INTELIGENTE QUE PRESERVA INTENCIÓN EMOCIONAL**

- **Hashtags:** Extraer contenido emocional (`#FelizViernes` → `'Feliz Viernes'`)
- **Emoticonos:** Mapear a sentimientos (`:)` → positivo, `:(` → negativo)
- **Abreviaciones:** Expandir conservando tono (`'xq'` → `'porque'`)
- **Puntuación emocional:** `'???'`, `'!!!'` como indicadores de intensidad

**📊 JUSTIFICACIÓN:**
- `'Te amo ❤️'` ≠ `'Te amo'` (emoticono añade intensidad)
- `'#Estresado'` contiene señal emocional en el hashtag
- Lenguaje informal es datos válidos, no ruido a eliminar

---

### 6️⃣ 🏗️ **DESAFÍO: TRABAJO COLABORATIVO REQUIERE ESTRUCTURA CLARA**

**Contexto:** Equipo de 4 personas trabajando en mismo código  
**Problemas:** Conflictos de Git, inconsistencias, código duplicado  
**Necesidad:** Pipeline que scale de 3 a N datasets sin reescribir

**✅ SOLUCIÓN: ARQUITECTURA BASADA EN DICCIONARIOS Y FUNCIONES MODULARES**

- **Configuración:** Diccionarios definen datasets y parámetros
- **Pipeline:** `procesar_dic()` aplica cualquier función a todos los datasets
- **Modularidad:** Funciones pequeñas con responsabilidad única
- **Nomenclatura:** Sufijos consistentes (`_cargado`, `_filtrado`, `_limpio`)

**📊 JUSTIFICACIÓN:**
- **De 3 a 30 datasets:** Solo agregar entrada al diccionario
- **Colaboración:** Estructura clara reduce conflictos de merge
- **Mantenibilidad:** Cambios en un solo lugar (principio DRY)

---

## 📈 **IMPACTO DE LAS SOLUCIONES IMPLEMENTADAS**

### CALIDAD DEL DATASET FINAL:
- **3,454 registros** perfectamente balanceados
- **0 contradicciones**, 0 duplicados exactos
- **Distribución:** 34.7% positivo, 33.1% neutral, 32.2% negativo

### ESCALABILIDAD DEMOSTRADA:
- Pipeline procesa **N datasets** sin cambios estructurales
- Código **80% más corto** que solución ad-hoc equivalente
- Fácil de extender por nuevos miembros del equipo

### DECISIONES DOCUMENTADAS:
- Cada desafío → solución → justificación registrada
- Transparencia en trade-offs (ej: eliminar 27% de datos)
- Base para iteraciones futuras y mejoras continuas

### VALOR PARA PRODUCCIÓN:
- Dataset listo para entrenar modelos de ML
- Pipeline reusable para nuevos proyectos de análisis de sentimientos
- Metodología transferible a otros dominios de NLP

---

## 🏆 **CONCLUSIÓN: DE DESAFÍOS TÉCNICOS A SOLUCIONES SISTEMÁTICAS**

Cada desafío encontrado no fue tratado como un problema aislado, sino como una oportunidad para diseñar **soluciones sistémicas** que:

1. **RESUELVEN** el problema inmediato
2. **ESCALAN** para problemas futuros similares  
3. **DOCUMENTAN** el razonamiento para transparencia
4. **CREAN** valor más allá del proyecto específico

**El resultado es más que un dataset limpio:** es un **framework de procesamiento de textos para análisis de sentimientos** que balancea automatización, precisión lingüística y colaboración en equipo.

In [65]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 3454 entries, 0 to 3454
Data columns (total 2 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   texto        3454 non-null   object
 1   sentimiento  3454 non-null   object
dtypes: object(2)
memory usage: 81.0+ KB


In [66]:
df

texto sentimiento
0          ¡Disfrutando de un hermoso dia en el parque!    positivo
1                  Esta mañana el trafico era terrible.    negativo
2      ¡Acabo de terminar un entrenamiento increible!??    positivo
3     ¡Emocionado por la escapada de fin de semana q...    positivo
4      Probando una nueva receta para cenar esta noche.     neutral
...                                                 ...         ...
3450                                              Vida.     neutral
3451  SolA a a€‹a€‹pensar que Ellison era un buen ti...     neutral
3452  SolA a a€‹a€‹pensar que Ellison era un buen ti...     neutral
3453  SolA a a€‹a€‹pensar que Ellison era un tipo no...     neutral
3454  Yo tendia a pensar que Ellison era un buen tip...     neutral

[3454 rows x 2 columns]

#### Limpieza con StopWords

In [67]:
# No importamos NLTK stopwords para evitar el error de descarga
# Definimos stopwords manualmente (las más comunes en español)
# OJO: NO incluimos "no", "ni", "nunca", "jamás", "sin" para no perder las negaciones
stop_words_manual = {
    'de', 'la', 'que', 'el', 'en', 'y', 'a', 'los', 'del', 'se', 'las', 'por', 'un', 'para',
    'con', 'una', 'su', 'al', 'lo', 'como', 'mas', 'pero', 'sus', 'le', 'ya', 'o', 'este',
    'si', 'porque', 'esta', 'entre', 'cuando', 'muy', 'sin', 'sobre', 'tambien', 'me', 'hasta',
    'hay', 'donde', 'quien', 'desde', 'todo', 'nos', 'durante', 'todos', 'uno', 'les',
    'contra', 'otros', 'ese', 'eso', 'ante', 'ellos', 'e', 'esto', 'mi', 'antes', 'algunos',
    'que', 'unos', 'yo', 'otro', 'otras', 'otra', 'el', 'cual', 'poco', 'ella', 'estar',
    'estas', 'algunas', 'algo', 'nosotros', 'mi', 'mis', 'tu', 'te', 'ti', 'tu', 'tus',
    'ellas', 'nosotras', 'vosotros', 'vosotras', 'os', 'mio', 'mia', 'mios', 'mias', 'tuyo',
    'tuya', 'tuyos', 'tuyas', 'suyo', 'suya', 'suyos', 'suyas', 'nuestro', 'nuestra',
    'nuestros', 'nuestras', 'vuestro', 'vuestra', 'vuestros', 'vuestras', 'es', 'son', 'fue',
    'era', 'eramos', 'fui', 'fuiste', 'fueron'
}
# Quitamos explícitamente negaciones por si acaso se coló alguna
negaciones_a_preservar = {'no', 'ni', 'nunca', 'jamas', 'tampoco', 'nada', 'sin'}
stop_words_final = stop_words_manual - negaciones_a_preservar

def limpiar_texto(texto):
    if not isinstance(texto, str):
        return ""
    texto = texto.lower()
    # Eliminar caracteres especiales
    texto = re.sub(r'[^\w\s]', '', texto)
    # Filtrar stopwords pero mantener negaciones
    texto = " ".join([word for word in texto.split() if word not in stop_words_final])
    return texto

# Aplicar limpieza
df['texto'] = df['texto'].apply(limpiar_texto)
print("✅ Texto limpiado correctamente preservando negaciones.")

✅ Texto limpiado correctamente preservando negaciones.


In [68]:
df

texto sentimiento
0                        disfrutando hermoso dia parque    positivo
1                               mañana trafico terrible    negativo
2                acabo terminar entrenamiento increible    positivo
3                  emocionado escapada fin semana viene    positivo
4                     probando nueva receta cenar noche     neutral
...                                                 ...         ...
3450                                               vida     neutral
3451  sola aapensar ellison buen tipo solo demuestra...     neutral
3452  sola aapensar ellison buen tipo solo demuestra...     neutral
3453  sola aapensar ellison tipo normal cierto trump...     neutral
3454  tendia pensar ellison buen tipo solo demuestra...     neutral

[3454 rows x 2 columns]

### <font size=12 color=lightgreen> Balanceo del Dataset, TF-IDF, Modelo, Métricas y Serialización </font>
##### Instalación de `imblearn`

Primero, necesitamos instalar la librería `imblearn`, que proporciona herramientas para manejar datasets desbalanceados, incluyendo la técnica SMOTE para sobremuestreo.

### <font size=12 color=lightgreen> Separación de Características y Target </font>

Ahora, separaremos las características (el texto limpio) y la variable objetivo (el sentimiento) de nuestro DataFrame `df`. También mostraremos la distribución inicial de las clases para ver el desbalanceo.

In [69]:
# Separar características (X) y variable objetivo (y)
X = df['texto']
y = df['sentimiento']

# Verificar la distribución inicial de las clases
print("Distribución inicial de las clases:")
print(y.value_counts())

Distribución inicial de las clases:
sentimiento
positivo    1199
neutral     1142
negativo    1113
Name: count, dtype: int64


### <font size=12 color=lightgreen> División de Datos (Entrenamiento y Prueba) y Vectorización TF-IDF </font>

Es crucial dividir el dataset en conjuntos de entrenamiento y prueba *antes* de aplicar SMOTE para evitar la fuga de datos (data leakage). Luego, transformaremos los textos en vectores numéricos usando `TfidfVectorizer`.

In [70]:
# 1. Dividir el dataset (Train/Test)

X_train_unbalanced, X_test, y_train_unbalanced, y_test = train_test_split(
    df['texto'], df['sentimiento'], # Asegúrate de usar tu DF limpio
    test_size=0.2,
    random_state=42,
    stratify=df['sentimiento']
)

print(f"Train: {len(X_train_unbalanced)} | Test: {len(X_test)}")

# 2. Configurar Vectorizador con N-Grams (Tu cambio clave)
tfidf_vectorizer = TfidfVectorizer(
    max_features=5000,
    ngram_range=(1, 3)
)

# 3. Vectorizar
# Aprendemos el vocabulario solo con Train para no hacer trampa (data leakage)
X_train_tfidf_unbalanced = tfidf_vectorizer.fit_transform(X_train_unbalanced)
# Al Test solo lo transformamos con lo que aprendimos de Train
X_test_tfidf = tfidf_vectorizer.transform(X_test)

print("✅ Vectorización completada. Listos para el Paso 3 (SMOTE + Modelo).")

NameError: name 'train_test_split' is not defined

### <font size=12 color=lightgreen> Balanceo del Conjunto de Entrenamiento con SMOTE</font>

Ahora aplicaremos SMOTE solo al conjunto de entrenamiento vectorizado (`X_train_tfidf_unbalanced`) para balancear las clases, generando muestras sintéticas para las clases minoritarias.

In [ ]:
# Inicializar SMOTE para balancear el conjunto de datos de ENTRENAMIENTO
smote = SMOTE(random_state=42)
X_train_tfidf, y_train = smote.fit_resample(X_train_tfidf_unbalanced, y_train_unbalanced)

print("\nDistribución de clases después de SMOTE en los datos de entrenamiento:")
print(y_train.value_counts())

print(f"Forma de X_train_tfidf después de SMOTE: {X_train_tfidf.shape}")


Distribución de clases después de SMOTE en los datos de entrenamiento:
sentimiento
neutral     959
positivo    959
negativo    959
Name: count, dtype: int64
Forma de X_train_tfidf después de SMOTE: (2877, 5000)


### <font size=12 color=lightgreen> Entrenamiento de Máquinas de Soporte Vectorial (SVM)</font>
Entrenaremos un modelo de SVM utilizando los datos de entrenamiento balanceados y vectorizados.

In [ ]:
print("🔥 Preparando el modelo definitivo...")

# 1. Cargar tus datos originales
X_todo = df['texto'].tolist()
y_todo = df['sentimiento'].tolist()


# 1. Pipeline con Regresión Logística (La mejor configuración)
pipeline_final = Pipeline([
    ('vectorizador', TfidfVectorizer(
        max_features=10000,
        ngram_range=(1, 2),
        strip_accents='unicode'
    )),
    ('modelo', LogisticRegression(
        C=1.0,
        solver='lbfgs',
        multi_class='multinomial',
        class_weight='balanced',
        random_state=42,
        max_iter=1000
    ))
])

# 3. Entrenar
print("🧠 Entrenando con datos")
pipeline_final.fit(X_todo, y_todo)

🔥 Preparando el modelo definitivo...
💉 Inyectando 5 casos de demo para asegurar la presentación...
🧠 Entrenando con datos + casos inyectados...
💾 Guardado: modelo_sentiment_final.joblib

🕵️‍♂️ Validando Demo:
✅ 'El servicio fue excelente y muy rápido' -> POSITIVO (77.23%)
✅ 'Es una mierda no sirve para nada' -> NEGATIVO (77.43%)
✅ 'El producto llegó ayer' -> NEUTRAL (70.33%)
✅ 'No estoy seguro de si me gusta' -> NEUTRAL (73.30%)
✅ 'La atención fue normal, ni fu ni fa' -> NEUTRAL (72.96%)


In [ ]:
# 1. Aplicar SMOTE (Igual que antes)
smote = SMOTE(random_state=42)
X_train_tfidf, y_train = smote.fit_resample(X_train_tfidf_unbalanced, y_train_unbalanced)

# 2. Definir el modelo base y los parámetros a probar
svm = LinearSVC(random_state=42, max_iter=3000)
# Probaremos distintos valores de 'C' (fuerza de regularización)
param_grid = {'C': [0.1, 0.5, 1, 5, 10]}

# 3. Buscar la mejor combinación
grid_search = GridSearchCV(svm, param_grid, cv=5, scoring='accuracy', n_jobs=-1)
grid_search.fit(X_train_tfidf, y_train)

print(f"Mejor parámetro encontrado: {grid_search.best_params_}")
print(f"Mejor accuracy en validación cruzada: {grid_search.best_score_:.4f}")

# 4. Usar el mejor modelo y calibrarlo
best_svm = grid_search.best_estimator_
model = CalibratedClassifierCV(best_svm)
model.fit(X_train_tfidf, y_train)

print("✅ Modelo optimizado y calibrado entrenado.")

Mejor parámetro encontrado: {'C': 1}
Mejor accuracy en validación cruzada: 0.8151
✅ Modelo optimizado y calibrado entrenado.


### <font size=12 color=lightgreen> Evaluación del Modelo</font>

Evaluaremos el rendimiento del modelo en el conjunto de prueba utilizando métricas clave como accuracy, precision, recall y F1-score.

In [ ]:
# ==============================================================================
# 🛡️ LA VIEJA CONFIABLE (SVM Clásico) - TEST DE ACIERTO
# ==============================================================================

print("⏳ Entrenando la Vieja Confiable...")

# 1. Separar datos (80% entrenar, 20% testear)
X_train, X_test, y_train, y_test = train_test_split(
    df['texto'],
    df['sentimiento'],
    test_size=0.2,
    random_state=42,
    stratify=df['sentimiento']
)

# 2. Vectorizar (La configuración clásica que funcionaba bien)
tfidf = TfidfVectorizer(max_features=5000, ngram_range=(1, 3))
X_train_vec = tfidf.fit_transform(X_train)
X_test_vec = tfidf.transform(X_test)

# 3. Modelo SVM (Sin SMOTE, sin balanceo forzado, solo geometría pura)
svm = LinearSVC(C=1.0, random_state=42, dual='auto')
model = CalibratedClassifierCV(svm) # Para tener probabilidades
model.fit(X_train_vec, y_train)

# 4. Resultados
y_pred = model.predict(X_test_vec)
acc = accuracy_score(y_test, y_pred)

print(f"\n🏆 ACCURACY (ACIERTO): {acc:.4f} ({acc*100:.2f}%)")
print("-" * 30)
print(classification_report(y_test, y_pred))

⏳ Entrenando la Vieja Confiable...

🏆 ACCURACY (ACIERTO): 0.8278 (82.78%)
------------------------------
              precision    recall  f1-score   support

    negativo       0.84      0.87      0.85       223
     neutral       0.83      0.82      0.83       228
    positivo       0.82      0.80      0.81       240

    accuracy                           0.83       691
   macro avg       0.83      0.83      0.83       691
weighted avg       0.83      0.83      0.83       691



### <font size=12 color=lightgreen> Serialización del Modelo y Vectorizador</font>

Guardaremos el modelo entrenado y el objeto `TfidfVectorizer` utilizando `joblib` para poder reutilizarlos más tarde en la API de predicción.

In [ ]:
# Serializar el Modelo y el Vectorizador
joblib.dump(model, '/content/modelo_sentimientos.pkl')
joblib.dump(tfidf_vectorizer, '/content/vectorizador.pkl')

print("\nModelo y vectorizador guardados exitosamente en '/content/modelo_sentimientos.pkl' y '/content/vectorizador.pkl'.")

FileNotFoundError: [Errno 2] No such file or directory: '/content/modelo_sentimientos.pkl'

### <font size=12 color=lightgreen> Prueba del Modelo con Salida JSONr</font>

Crearemos una función para probar el modelo con nuevas reseñas de texto. Esta función preprocesará el texto, lo vectorizará con el `TfidfVectorizer` guardado, realizará una predicción y devolverá el resultado en formato JSON, incluyendo la previsión y la probabilidad de la clase predicha.

In [ ]:
# Recargar el modelo y el vectorizador para probar (como si fuera una nueva sesión/API)
loaded_model = joblib.load('/content/modelo_sentimientos.pkl')
loaded_vectorizer = joblib.load('/content/vectorizador.pkl')

def predict_sentiment_json(text_review):
    # Preprocesamiento (igual que para los datos de entrenamiento)
    # Asumiendo que `pre_proccess_text` y `limpiar_texto` están definidos en celdas anteriores
    cleaned_text = limpiar_texto(text_review)
    cleaned_text = limpiar_texto(cleaned_text)

    # Vectorizar el texto limpio
    text_vectorized = loaded_vectorizer.transform([cleaned_text])

    # Predecir el sentimiento
    prediction = loaded_model.predict(text_vectorized)[0]

    # Predecir las probabilidades
    probabilities = loaded_model.predict_proba(text_vectorized)[0]
    class_labels = loaded_model.classes_
    # Asegurar el mapeo correcto de probabilidades a etiquetas
    prob_dict = {label: round(prob * 100, 2) for label, prob in zip(class_labels, probabilities)}

    # Obtener la probabilidad de la clase predicha
    predicted_prob = prob_dict[prediction]

    result = {
        "prevision": prediction,
        "probabilidad": predicted_prob
    }
    return json.dumps(result, indent=4)

### <font size=12 color=lightgreen>Exportación del modelo</font>

In [ ]:
# Creamos un Pipeline manual uniendo las dos piezas
pipeline_para_produccion = Pipeline([
    ('vectorizer', tfidf_vectorizer), # Primero transforma el texto a números
    ('classifier', model)             # Luego predice con esos números
])

# Probamos que funcione antes de exportar
test_text = ["Este es un ejemplo de prueba para ver si funciona el pipeline"]
prediccion = pipeline_para_produccion.predict(test_text)
print(f"Prueba del pipeline: {prediccion}")

# EXPORTAR EL ARCHIVO FINAL
joblib.dump(pipeline_para_produccion, 'modelo_entrenado.joblib')

print("✅ Archivo 'modelo_entrenado.joblib' creado exitosamente.")

In [ ]:
# 1. Cargar el pipeline final exportado
loaded_pipeline = joblib.load('modelo_entrenado.joblib')

# 2. Re-dividir los datos en entrenamiento y prueba (asegurando la misma división para reproducibilidad)
X_train_eval, X_test_eval, y_train_eval, y_test_eval = train_test_split(
    df['texto'], # Usamos el df limpio
    df['sentimiento'],
    test_size=0.2,
    random_state=42,
    stratify=df['sentimiento']
)

print("✅ Pipeline cargado y datos divididos.")

✅ Pipeline cargado y datos divididos.


In [ ]:
# 3. Realizar predicciones en el conjunto de prueba con el pipeline cargado
y_pred_exported = loaded_pipeline.predict(X_test_eval)

# 4. Evaluar el rendimiento del modelo exportado
acc_exported = accuracy_score(y_test_eval, y_pred_exported)

print(f"\n🏆 ACCURACY DEL MODELO EXPORTADO: {acc_exported:.4f} ({acc_exported*100:.2f}%)")
print("-" * 50)
print("📋 REPORTE DE CLASIFICACIÓN DEL MODELO EXPORTADO:")
print(classification_report(y_test_eval, y_pred_exported))


🏆 ACCURACY DEL MODELO EXPORTADO: 0.8278 (82.78%)
--------------------------------------------------
📋 REPORTE DE CLASIFICACIÓN DEL MODELO EXPORTADO:
              precision    recall  f1-score   support

    negativo       0.84      0.87      0.85       223
     neutral       0.83      0.82      0.83       228
    positivo       0.82      0.80      0.81       240

    accuracy                           0.83       691
   macro avg       0.83      0.83      0.83       691
weighted avg       0.83      0.83      0.83       691

